db/python/tables/family.py

```diff
@@ -29,17 +29,39 @@ async def get_projects_by_family_ids(self, family_ids: List[int]) -> Set[Project
             )
         return projects
 
-    async def get_families(self, project: int = None) -> List[Family]:
+    async def get_families(
+        self, project: int = None, participant_ids: List[int] = None
+    ) -> List[Family]:
         """Get all families for some project"""
-        _query = """\
-SELECT id, external_id, description, coded_phenotype, project
-FROM family
-WHERE project = :project"""
+        _query = """
+            SELECT id, external_id, description, coded_phenotype, project
+            FROM family
+        """
 
-        rows = await self.connection.fetch_all(
-            _query, {'project': project or self.project}
-        )
-        families = [Family.from_db(dict(r)) for r in rows]
+        values: Dict[str, Any] = {'project': project or self.project}
+        where: List[str] = []
+
+        if participant_ids:
+            _query += """
+                JOIN family_participant
+                ON family.id = family_participant.family_id
+            """
+            where.append(f'participant_id IN :pids')
+            values['pids'] = participant_ids
+
+        if project or self.project:
+            where.append('project = :project')
+
+        if where:
+            _query += 'WHERE ' + ' AND '.join(where)
+
+        rows = await self.connection.fetch_all(_query, values)
+        seen = set()
+        families = []
+        for r in rows:
+            if r.id not in seen:
+                families.append(Family.from_db(dict(r)))
+                seen.add(r.id)
```

Comment:
This works, but I think it would be cleaner to make the database do this by using a GROUP BY (on all values returned), eg:

```sql
SELECT f.id, f.external_id, f.description, f.coded_phenotype, f.project
FROM family f
INNER JOIN family_participant fp
   ON f.id = fp.family_id             
WHERE fp.participant_id IN (97, 100, 101) AND f.project = 3
GROUP BY f.id, f.external_id, f.description, f.coded_phenotype, f.project
```

I'd probably encourage writing separate queries for when the participants are specified, and when they're not. Gives you more flexibility to add the GROUP BY.

Reply:
Yeah I was thinking about switching to the native group by. But ended up with weird issues since if no pariticpants or samples or specified there is no group by and the aggregate was acting weird.